backend/utils/pubsub.js

```diff
@@ -0,0 +1,261 @@
+/* globals logger */
+'use strict'
+
+import {
+  NOTIFICATION_TYPE,
+  REQUEST_TYPE,
+  RESPONSE_TYPE,
+  createClient,
+  createMessage,
+  messageParser
+} from '~/shared/pubsub.js'
+
+import type { Message, SubMessage, UnsubMessage } from '~/shared/pubsub.js'
+import type { JSONType } from '~/shared/types.js'
+
+const { bold } = require('chalk')
+const WebSocket = require('ws')
+
+// ====== API ====== //
+
+const { PUB, SUB, UNSUB } = NOTIFICATION_TYPE
+const { ERROR, SUCCESS } = RESPONSE_TYPE
+
+// Re-export some useful things from the shared module.
+export { createClient, createMessage, NOTIFICATION_TYPE, REQUEST_TYPE, RESPONSE_TYPE }
+
+export function createNotification (
+  type: string,
+  data: {| contractID: string, socketID: number |}
+): string {
+  return JSON.stringify({ type, ...data })
+}
+
+export function createResponse (type: string, data: JSONType): string {
+  return JSON.stringify({ type, data })
+}
+
+/**
+ * Creates a pubsub server instance.
+ *
+ * @param {(http.Server|https.Server)} server - A Node.js HTTP/S server to attach to.
+ * @param {Object?} options
+ * {object?} clientHandlers - Custom handlers for socket events.
+ * {object?} messageHandlers - Custom handlers for different message types.
+ * {object?} serverHandlers - Custom handlers for server events.
+ *
+ * {number} backlog - The maximum length of the queue of pending connections.
+ * {boolean} clientTracking - Specifies whether or not to track clients.
+ * {Function} handleProtocols - A function which can be used to handle the WebSocket subprotocols.
+ * {number} maxPayload - The maximum allowed message size in bytes.
+ * {string} path - Accept only connections matching this path.
+ * {(boolean|object)} perMessageDeflate - Enables/disables per-message deflate.
+ * @returns {Object}
+ */
+export function createServer (httpServer: Object, options?: Object = {}): Object {
+  const server = new WebSocket.Server({
+    ...defaultOptions,
+    ...options,
+    server: httpServer
+  })
+  server.customClientHandlers = { ...options.clientHandlers }
+  server.customServerHandlers = { ...options.serverHandlers }
```

Comment:
To remove any coupling between them and the passed in `options` object.

This way, subsequent changes to the `options` object cannot alter server behavior as a side effect.

Even in nobody ever mutates the `options` object after using it, decoupling could still help the engine and various tools.

Reply:
Is there a concrete reason to believe that that will occur here, or is this intended as a general precaution? In the latter case, it doesn't seem like this is currently done anywhere else in the code, from a quick search - so perhaps this should be checked for other code as well, then. Otherwise you still can't assume that options passed to an arbitrary function are 'locked in' at call time.

That having been said, having this as a general practice is likely to turn into a rabbit hole, if for whatever reason we end up working with nested data structures in options at some point. Deep copies are expensive and difficult to express sensibly, especially when you're dealing with structures that contain more than just arrays and object literals.

Deep nested structures are frequently also those most subject to mutation (because of aforementioned deep-copy cost), so you'd end up with a situation where you can safely assume that inputs are not mutated afterwards... *except* for those structures which are actually *likely* to get mutated.

Because of this, I've stopped doing this in my own projects - it turned out to just not be worth it in practice for me. Instead, I just handle data immutably wherever possible (so inputs are super unlikely to ever mutate at all), and carefully reason through unavoidable mutation, which has more or less the same result.

Reply:
It is only intended as a general practice.

You have a point that it doesn't prevent nested options to stay coupled. That's one reason why this implementation has only flat options, whereas Primus accepts nested options such as `recovery` and `strategy`.

Thank you for sharing your experience, that's interesting. Plain assignment can certainly be used if you think copying does more harm than good.

Reply:
So I have followed your advice, however I also had to append `|| {}` instead of just using a simple property assignment.

Now we have `options.clientHandlers || {}` instead of `{ ...options.clientHandlers }`. Both have downsides IMHO.